### DISCUSSION TITLE
how to call outer variable

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1517] [Q&A]
i want to define a custom function block at my st program decl,this custom function block  is not define at st file,but get this from another way like a json file 
![image](https://github.com/eclipse-langium/langium/assets/26995040/b6a38c1e-af7b-4539-af24-5b0fa909a8ff)
for var declaration at st program ,it seems easy to achieve by changing my langium rule,but if i want to call it inner vaiable  lke below 
![image](https://github.com/eclipse-langium/langium/assets/26995040/c1535b73-fca8-46dc-95c6-c825951a3daf)
i can not achieve it by override DefaultScopeProvider$getScope

**Comment by cdietrich** [comment]
please check out https://langium.org/docs/recipes/builtin-library/ https://github.com/eclipse-langium/langium/discussions/1491

**Comment by msujew** [comment]
Hey @zrwsmd,

we have had this question multiple times in the forum already. Please refer to https://github.com/eclipse-langium/langium/discussions/984 for a longer discussion on the topic and potential solutions.
